Framework/Geometry/src/Objects/CSGObject.cpp

```diff
@@ -1142,11 +1166,50 @@ TrackDirection CSGObject::calcValidType(const Kernel::V3D &point, const Kernel::
 }
 
 /**
- * Find soild angle of object wrt the observer. This interface routine calls
- * either
- * getTriangleSoldiAngle or getRayTraceSolidAngle. Choice made on number of
- * triangles
- * in the discete surface representation.
+ * Check if an intercept is guiding the ray into the shape or leaving the shape
+ * @param prePt :: the previous point on the Line
+ * @param curPt :: the current point on the Line
+ * @param nxtPt :: Unit vector of the track
+ * @retval 1 :: Entry point
+ * @retval -1 :: Exit Point
+ * @retval 0 :: Not valid / double valid
+ */
+TrackDirection CSGObject::calcValidTypeBy3Points(const Kernel::V3D &prePt, const Kernel::V3D &curPt,
+                                                 const Kernel::V3D &nxtPt) const {
+  // upstream point
+  const auto upstreamPt = (prePt + curPt) * 0.5;
+  const int upstreamPtInsideShape = isValid(upstreamPt);
```

Comment:
`CSGObject::isValid(const V3D &)` returns a `bool` rather than an `int`. This section of code would be clearer without the implicit cast.

Reply:
good point, I will switch to `const auto` to use the default return type.